src/main/kotlin/org/rust/ide/inspections/RsNeedlessLifetimesInspection.kt

```diff
@@ -0,0 +1,240 @@
+ */
+class RsNeedlessLifetimesInspection : RsLocalInspectionTool() {
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): RsVisitor = object : RsVisitor() {
+        override fun visitFunction(fn: RsFunction) {
+            if (hasWhereLifetimes(fn.whereClause)) return
+
+            val boundsLifetimes = fn.typeParameters
+                .flatMap { it.bounds }
+                .map { it.bound }
+                .flatMap { bound ->
+                    val collector = LifetimesCollector()
+                    bound.accept(collector)
+                    if (collector.lifetimes.any { it is ReferenceLifetime.Named }) return
+
+                    val lifetimes = bound.traitRef?.path?.typeArgumentList?.lifetimeList.orEmpty()
+                    if (lifetimes.any { !it.isStatic && !it.isElided }) return
+                    lifetimes
+                }
+            if (couldUseElision(fn, boundsLifetimes)) {
+                holder.registerProblem(
+                    fn,
+                    "Explicit lifetimes given in parameter types where they could be elided",
+                    ProblemHighlightType.WEAK_WARNING,
+                    TextRange(
+                        fn.fn.startOffsetInParent,
+                        fn.block?.getPrevNonCommentSibling()?.endOffsetInParent ?: fn.identifier.endOffsetInParent
+                    )
+                )
+            }
+        }
+    }
+}
+
+/**
+ * There are two scenarios where elision works:
+ * - no output references, all input references have different LT
+ * - output references, exactly one input reference with same LT
+ * All lifetimes must be unnamed, 'static or defined without bounds on the level of the current item.
+ */
+private fun couldUseElision(fn: RsFunction, boundsLifetimes: List<RsLifetime>): Boolean {
+    // check named LTs
+    val allowedLifetimes = allowedLifetimesFrom(fn.lifetimeParameters)
+
+    // these will collect all the lifetimes for references in arg/return types
+    val inputCollector = LifetimesCollector()
+    val outputCollector = LifetimesCollector()
+
+    // extract lifetimes in input argument types
+    fn.valueParameterList?.let { param ->
+        param.selfParameter?.accept(inputCollector)
+        param.valueParameterList.forEach { it.typeReference?.accept(inputCollector) }
+    }
+    if (inputCollector.abort) return false
+
+    // extract lifetimes in output type
+    fn.retType?.typeReference?.accept(outputCollector)
+    if (outputCollector.abort) return false
+
+    val inputLifetimes = inputCollector.lifetimes + lifetimesFromBounds(boundsLifetimes)
+    val outputLifetimes = outputCollector.lifetimes
+
+    if (BodyLifetimeChecker.checkLifetimesUsedInBody(fn.block)) return false
+
+    // check for lifetimes from higher scopes
+    if (inputLifetimes.chain(outputLifetimes).any { !allowedLifetimes.contains(it) }) return false
+
+    // no input lifetimes? easy case!
+    if (inputLifetimes.isEmpty()) return false
```

Comment:
I'd move empty checking just below `inputLifetimes` and `outputLifetimes` declarations.